### PR TITLE
fix for dotnet hooks with prefixes

### LINKED ISSUE #2629: dotnet install fails for prefixed packages
### search you tried in the issue tracker

dotnet tool

### describe your issue

A bit of an oversight when constructing `tool_name` here:

https://github.com/pre-commit/pre-commit/blob/cb0bcfd67fc35e91f7b2eca7e33bceda459dca77/pre_commit/languages/dotnet.py#L60-L63

E.g.

```console
$ pre-commit try-repo https://github.com/rkm/sample-dotnet-tool
[INFO] Initializing environment for https://github.com/rkm/sample-dotnet-tool.
===============================================================================
Using config:
===============================================================================
repos:
-   repo: https://github.com/rkm/sample-dotnet-tool
    rev: e53a3601bc06bb038dac30da813572291dd8d58f
    hooks:
    -   id: sample-dotnet-tool
===============================================================================
[INFO] Installing environment for https://github.com/rkm/sample-dotnet-tool.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
An unexpected error has occurred: CalledProcessError: command: ('/home/rkm/bin/dotnet', 'tool', 'install', '--tool-path', '/tmp/tmp6bk4v26x/repotefhurdg/dotnetenv-default/bin', '--add-source', 'pre-commit-build', 'Rkm')
return code: 1
expected return code: 0
stdout:
    /tmp/1873db78-d0a7-48ba-bbff-10a7ef85a2a6/restore.csproj : error NU1101: Unable to find package rkm. No packages exist with this id in source(s): /tmp/tmp6bk4v26x/repotefhurdg/pre-commit-build, nuget.org

stderr:
    The tool package could not be restored.
    Tool 'rkm' failed to install. This failure may have been caused by:

    * You are attempting to install a preview release and did not use the --version option to specify the version.
    * A package by this name was found, but it was not a .NET tool.
    * The required NuGet feed cannot be accessed, perhaps because of an Internet connection problem.
    * You mistyped the name of the tool.

    For more reasons, including package naming enforcement, visit https://aka.ms/failure-installing-tool

Check the log at /home/rkm/.cache/pre-commit/pre-commit.log
```

### pre-commit --version

pre-commit 2.20.0

### .pre-commit-config.yaml

```yaml
repos:
-   repo: https://github.com/rkm/sample-dotnet-tool
    rev: e53a3601bc06bb038dac30da813572291dd8d58f
    hooks:
    -   id: sample-dotnet-tool
```

### ~/.cache/pre-commit/pre-commit.log (if present)

_No response_

### PR DESCRIPTION
**Pull request by rkm**
(Recreated from #2630)

Fixes #2629

### CODE REVIEW

File: pre_commit/languages/dotnet.py [hunk 1]
```diff
@@ -58,9 +59,15 @@ def install_environment(
...
+        # https://learn.microsoft.com/en-us/nuget/concepts/package-versioning#version-basics
+        tool_name_re = re.compile(r'(.*)\.\d+\.\d+\.\d+(-.*)?\.nupkg')
+
         build_outputs = os.listdir(os.path.join(prefix.prefix_dir, build_dir))
         for output in build_outputs:
-            tool_name = output.split('.')[0]
+
+            tool_name_match = tool_name_re.match(output)
+            assert tool_name_match
+            tool_name = tool_name_match.group(1)
```
Comment (asottile) [inline pre_commit/languages/dotnet.py:70]: this seems fragile -- can we read the name out of the metadata itself? or install the `.nupkg` file directly?

Comment (rkm) [reply pre_commit/languages/dotnet.py]: I've committed a version which reads the xml from the nupkg (zip file), but it needs tidied up a bit

Comment (rkm) [reply pre_commit/languages/dotnet.py]: There might be a way to simplify it, but this version seems to work.

**Review by asottile — APPROVED** [review]

**Pull request merged**